### PR TITLE
TY&RES: fix impls selection for impl Trait

### PR DESCRIPTION
**Pull request by vlad20012**

### CODE REVIEW

File: src/main/kotlin/org/rust/lang/core/resolve/indexes/RsImplIndex.kt [hunk 1]
```diff
@@ -39,6 +39,13 @@ class RsImplIndex : AbstractStubIndex<TyFingerprint, RsImplItem>() {
             return (impls.asSequence() + freeImpls.asSequence()).filter { it.crateRoot != null }
         }
 
+        /** return impls for generic type `impl<T> Trait for T {}` */
+        fun findFreeImpls(project: Project): Sequence<RsImplItem> {
```
Comment (mchernyavsky) [inline src/main/kotlin/org/rust/lang/core/resolve/indexes/RsImplIndex.kt]: Can we use this function inside `findPotentialImpls`?

File: src/main/kotlin/org/rust/lang/core/resolve/ImplLookup.kt [hunk 2]
```diff
@@ -628,10 +639,10 @@ class ImplLookup(
         return when (selfTy) {
             is TyTypeParameter -> lookupAssocTypeInBounds(selfTy.getTraitBoundsTransitively(), res.impl, assocType)
             is TyTraitObject -> selfTy.trait.assoc[assocType]
-            is TyAnon -> lookupAssocTypeInBounds(selfTy.getTraitBoundsTransitively(), res.impl, assocType)
             else -> {
                 lookupAssocTypeInSelection(res, assocType)
                     ?: lookupAssocTypeInBounds(getHardcodedImpls(selfTy), res.impl, assocType)
+                    ?: (selfTy as? TyAnon)?.let { lookupAssocTypeInBounds(it.getTraitBoundsTransitively(), res.impl, assocType) }
```
Comment (mchernyavsky) [inline src/main/kotlin/org/rust/lang/core/resolve/ImplLookup.kt]: Can you cover this case with a test?

Comment (vlad20012) [reply src/main/kotlin/org/rust/lang/core/resolve/ImplLookup.kt]: I sure that it is covered

**Pull request merged**